tests/validators/test_dataclasses.py

```diff
@@ -1592,3 +1592,114 @@ def _wrap_validator(cls, v, validator, info):
     gc.collect()
 
     assert ref() is None
+
+
+init_test_cases = [
+    ({'a': 'hello', 'b': 'bye'}, 'ignore', {'a': 'hello', 'b': 'HELLO'}),
+    ({'a': 'hello'}, 'ignore', {'a': 'hello', 'b': 'HELLO'}),
+    ({'a': 'hello'}, 'allow', {'a': 'hello', 'b': 'HELLO'}),
+    (
+        {'a': 'hello', 'b': 'bye'},
+        'forbid',
+        Err(
+            'Unexpected keyword argument',
+            errors=[
+                {
+                    'type': 'unexpected_keyword_argument',
+                    'loc': ('b',),
+                    'msg': 'Unexpected keyword argument',
+                    'input': 'bye',
+                }
+            ],
+        ),
+    ),
+    ({'a': 'hello'}, 'forbid', {'a': 'hello', 'b': 'HELLO'}),
+]
+
+
+@pytest.mark.parametrize(
+    'input_value,extra_behavior,expected',
+    [
+        *init_test_cases,
+        ({'a': 'hello', 'b': 'bye'}, 'allow', {'a': 'hello', 'b': 'HELLO'}),
```

Comment:
I would add a comment above this line indicating that we are open to changing this behavior and have just included the test to document it. And mentioning as you did below that we plan to disallow this combination in pydantic for now.